ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/DB/HdWallet/Delete.hs

```diff
@@ -1,24 +1,66 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
 -- | DELETE operatiosn on HD wallets
 module Ariadne.Wallet.Cardano.Kernel.DB.HdWallet.Delete (
-    deleteHdRoot
+    DeleteHdAccountError(..)
+  , DeleteHdRootError(..)
+  , deleteHdRoot
   , deleteHdAccount
   ) where
 
 import Universum
 
-import Control.Lens (at, (.=))
-
+import Data.SafeCopy (base, deriveSafeCopySimple)
+import Control.Lens ((%=))
 import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
 import Ariadne.Wallet.Cardano.Kernel.DB.Util.AcidState
+import Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet
 
 {-------------------------------------------------------------------------------
   DELETE
 -------------------------------------------------------------------------------}
 
--- | Delete a wallet
-deleteHdRoot :: HdRootId -> Update' HdWallets e ()
-deleteHdRoot rootId = zoom hdWalletsRoots $ at rootId .= Nothing
+-- | Delete a wallet with the whole subtree (addresses and accounts). 
+deleteHdRoot :: HdRootId -> Update' HdWallets DeleteHdRootError ()
+deleteHdRoot rootId = do 
+  zoomHdRootId DeleteHdRootExists rootId $ return ()
+  hdWalletsAddresses %= deleteIxAll rootId 
+  hdWalletsAccounts  %= deleteIxAll rootId 
+  hdWalletsRoots     %= deleteIxAll rootId 
+
+-- | Delete an account with its addresses.
+deleteHdAccount :: HdAccountId -> Update' HdWallets DeleteHdAccountError ()
+deleteHdAccount accId = do
+  zoomHdAccountId DeleteHdAccountExists accId $ return ()
+  hdWalletsAddresses %= deleteIxAll accId 
+  hdWalletsAccounts  %= deleteIxAll accId 
+
+
+{-------------------------------------------------------------------------------
+  Errors
+-------------------------------------------------------------------------------}
+
+-- | Errors thrown by 'deleteHdWallet'
+data DeleteHdRootError =
+    -- | We don't have a wallet with the specified ID
+    DeleteHdRootExists UnknownHdRoot
```

Comment:
Why is it called `Exists` if the wallet **does not** exist? Same for `DeleteHdAccountExists`.